### PR TITLE
Fix false positives/negatives in selector-pseudo-class/element-no-unknown

### PR DESCRIPTION
**Pull request by Mouvedia**
followup of #5956
see https://github.com/stylelint/stylelint/pull/5956#discussion_r822233829

> Which issue, if any, is this issue related to?

#2264

> Is there anything in the PR that needs further explanation?

Ill add comments on the relevant lines.
Remember to check the commit's message.

### CODE REVIEW

File: lib/reference/keywordSets.js [hunk 1]
```diff
@@ -244,16 +295,17 @@ keywordSets.otherPseudoClasses = new Set([
 	'first-child',
 	'first-of-type',
 	'focus',
-	'focus-ring',
```
Comment (Mouvedia) [inline lib/reference/keywordSets.js:247]: from an old draft
cf https://caniuse.com/css-focus-visible

File: lib/reference/keywordSets.js [hunk 2]
```diff
@@ -272,35 +324,46 @@ keywordSets.otherPseudoClasses = new Set([
...
-	'scrollbar-track-piece',
-	'scrollbar-thumb',
-	'scrollbar-corner',
-	'resize',
-]);
-
-// TODO migrate to vendorSpecificPseudoClasses
-// see #2264
-keywordSets.webkitProprietaryPseudoClasses = new Set([
+// https://webkit.org/blog/363/styling-scrollbars/
```
Comment (Mouvedia) [inline lib/reference/keywordSets.js:365]: this source was the cause of the typo

File: lib/reference/keywordSets.js [hunk 3]
```diff
@@ -272,35 +324,46 @@ keywordSets.otherPseudoClasses = new Set([
 	'scope',
 	'state',
 	'target',
+	'unresolved',
 	'user-invalid',
 	'user-valid',
-	'invalid',
 	'valid',
 	'visited',
+	'window-inactive', // for ::selection (chrome)
```
Comment (Mouvedia) [inline lib/reference/keywordSets.js:332]: also in webkitProprietaryPseudoClasses

File: lib/reference/keywordSets.js [hunk 4]
```diff
@@ -244,16 +295,17 @@ keywordSets.otherPseudoClasses = new Set([
 	'first-child',
 	'first-of-type',
 	'focus',
-	'focus-ring',
 	'focus-within',
 	'focus-visible',
 	'fullscreen',
+	'fullscreen-ancestor',
```
Comment (Mouvedia) [inline lib/reference/keywordSets.js:301]: supported by Opera 12

File: lib/reference/keywordSets.js [hunk 5]
```diff
@@ -238,22 +289,22 @@ keywordSets.otherPseudoClasses = new Set([
 	'default',
 	'defined',
 	'disabled',
-	'drop',
```
Comment (Mouvedia) [inline lib/reference/keywordSets.js:241]: https://github.com/w3c/csswg-drafts/issues/2257#issuecomment-365675701

**Comment by Mouvedia** [comment]
There are many more that are missing but Ill stop here; the PR has to be reviewable.
e.g. `::-webkit-media-*` are missing

**Review by jeddy3 — CHANGES_REQUESTED** [review]
Thanks for all these!

Can you run `npm run format` to fix the linting, please?

**Comment by Mouvedia** [comment]
> Can you run `npm run format` to fix the linting, please?

If I did run `npm run format` it would add 50 files to the diff and wouldn't fix any issues in the files that I have added.
IMO there's a problem with the CI. It shouldn't run on _all_ files.

**Comment by ybiquitous** [comment]
When I run `npm run format`, there is one change as below:

```diff
diff --git a/lib/rules/selector-pseudo-class-no-unknown/index.js b/lib/rules/selector-pseudo-class-no-unknown/index.js
index 14aab769..3ae5f385 100644
--- a/lib/rules/selector-pseudo-class-no-unknown/index.js
+++ b/lib/rules/selector-pseudo-class-no-unknown/index.js
@@ -93,7 +93,9 @@ const rule = (primary, secondaryOptions) => {
 						let prevPseudoElement = pseudoNode;
 
 						do {
-							prevPseudoElement = /** @type {import('postcss-selector-parser').Base} */ (prevPseudoElement.prev());
+							prevPseudoElement = /** @type {import('postcss-selector-parser').Base} */ (
+								prevPseudoElement.prev()
+							);
 
 							if (prevPseudoElement && prevPseudoElement.value.slice(0, 2) === '::') {
 								break;
```

File: lib/rules/selector-pseudo-class-no-unknown/index.js [hunk 6]
```diff
@@ -17,7 +17,8 @@ const { isString } = require('../../utils/validateTypes');
 const ruleName = 'selector-pseudo-class-no-unknown';
 
 const messages = ruleMessages(ruleName, {
-	rejected: (selector) => `Unexpected unknown pseudo-class selector "${selector}"`,
+	rejected: (/** @type {any} */ selector) =>
+		`Unexpected unknown pseudo-class selector "${selector}"`,
```
Comment (ybiquitous) [inline lib/rules/selector-pseudo-class-no-unknown/index.js:21]: [suggest] This addition of `/** @type {any} */` seems needless. Is it possible to remove it?

Comment (Mouvedia) [reply lib/rules/selector-pseudo-class-no-unknown/index.js:21]: @ybiquitous I have a error without it.

```sh
lib/rules/selector-pseudo-element-no-unknown/index.js:17:13 - error TS7006: Parameter 'selector' implicitly has an 'any' type.

17  rejected: (selector) => `Unexpected unknown pseudo-element selector "${selector}"`,

Comment (ybiquitous) [reply lib/rules/selector-pseudo-class-no-unknown/index.js:21]: Can you try clean-installing `node_modules/` via `npm ci`?

**Comment by Mouvedia** [comment]
> When I run `npm run format`, there is one change as below:
> 
> ```diff
> diff --git a/lib/rules/selector-pseudo-class-no-unknown/index.js b/lib/rules/selector-pseudo-class-no-unknown/index.js
> index 14aab769..3ae5f385 100644
> --- a/lib/rules/selector-pseudo-class-no-unknown/index.js
> +++ b/lib/rules/selector-pseudo-class-no-unknown/index.js
> @@ -93,7 +93,9 @@ const rule = (primary, secondaryOptions) => {
>  						let prevPseudoElement = pseudoNode;
>  
>  						do {
> -							prevPseudoElement = /** @type {import('postcss-selector-parser').Base} */ (prevPseudoElement.prev());
> +							prevPseudoElement = /** @type {import('postcss-selector-parser').Base} */ (
> +								prevPseudoElement.prev()
> +							);
>  
>  							if (prevPseudoElement && prevPseudoElement.value.slice(0, 2) === '::') {
>  								break;
> ```

I had prettier version 2.2.1.
I installed 2.5.1 now it shows up.

File: lib/rules/selector-pseudo-class-no-unknown/index.js (hunk 6 above)
Comment (Mouvedia) [reply lib/rules/selector-pseudo-class-no-unknown/index.js]: Done.

**Review by ybiquitous — APPROVED** [review]
@Mouvedia Thank you! 👍🏼

**Comment by jeddy3** [comment]
> - Fixed: `selector-pseudo-*-no-unknown` false positives and negatives ([#5959](https://github.com/stylelint/stylelint/pull/5959)).

**Pull request merged**